File: android_native/docx/src/main/kotlin/com/afluffywaffle/layuv/docx/PlainTextMapper.kt
package com.afluffywaffle.layuv.docx

/** A bold and/or italic run over `[start, end)` char offsets into [PlainMap.plain]. */
data class FormatSpan(val start: Int, val end: Int, val bold: Boolean, val italic: Boolean)

/**
 * A heading paragraph for the document-outline navigator. [level] is 0-based
 * (0 = Heading 1). [charOffset] is the start of the heading's text in
 * [PlainMap.plain] — divide by `plain.length` for the 0.0–1.0 jump fraction,
 * the same coordinate system as annotation positions.
 */
data class Heading(val text: String, val level: Int, val charOffset: Int)

/**
 * A whole paragraph over `[start, end)` char offsets (end excludes the trailing
 * `\n`) styled with a Word paragraph border and/or paragraph-level shading —
 * e.g. a block-quote/callout paragraph (`<w:pBdr>` + `<w:shd>` inside `<w:pPr>`).
 * Distinct from run-level `<w:rPr><w:shd>` which imports as a `.highlight`
 * annotation instead (see NativeImport).
 */
data class ParagraphStyleSpan(val start: Int, val end: Int, val blockquote: Boolean)

/**
 * The canonical plain text [plain] plus, for each of its UTF-16 code units,
 * the char offset [xmlOffsets] into the source document.xml. The two arrays
 * are parallel: `xmlOffsets.size == plain.length`.
 *
 * [formats] is a READ-ONLY, additive overlay (direct `<w:b>`/`<w:i>` run
 * formatting) for display only — it does NOT affect [plain]/[xmlOffsets], so
 * anchoring and write-back round-trip exactly as before.
 */
class PlainMap(
    val plain: String,
    val xmlOffsets: IntArray,
    val formats: List<FormatSpan> = emptyList(),
    /** Heading paragraphs in document order — drives the navigation outline. */
    val headings: List<Heading> = emptyList(),
    /** Block-quote-styled paragraphs in document order — display only, same overlay contract as [formats]. */
    val paragraphStyles: List<ParagraphStyleSpan> = emptyList(),
)

/**
 * Builds the ONE canonical plain-text string the native reader renders AND
 * anchors annotations against (offsets into [PlainMap.plain] ARE the anchor
 * coordinate system).
 *
 * This is the CLEAN extraction — see [[native-android-port]]. It is
 * byte-identical to docx_store._buildPlainMap (plain + xmlOffsets) on ordinary
 * prose, but parses real element names so `<w:tab/>`, `<w:tbl>`, `<w:tr>`,
 * `<w:tc>` are no longer mis-read as `<w:t>` (the legacy regex bug), and it
 * decodes numeric character references.
 *
 * Extraction rules (mirror of `buildCleanMap` in
 * android_native/tools/golden_gen/gen_goldens.dart):
 * ```
 *   <w:t>…</w:t>      decoded text; one xmlOffset per code unit (base+i)
 *   <w:tab/>          '\t'   offset = tag start
 *   <w:br/> <w:cr/>   '\n'   offset = tag start
 *   </w:p>            '\n'   offset = tag start   (trailing newline kept)
 *   everything else   ignored
 * ```
 * Iteration is over UTF-16 code units (Kotlin `Char`), matching Dart's
 * `String[i]` indexing, so surrogate pairs (emoji) split identically and
 * offsets stay aligned with the Dart-generated goldens.
 */
object PlainTextMapper {

    /**
     * 1-indexed paragraph number containing char offset [offset] into [plain]. Exact —
     * paragraphs in [plain] are delimited one-for-one by the `\n` each `</w:p>` emits (see
     * the extraction rules above), so this is a plain newline count, not an approximation.
     */
    fun paragraphIndex(plain: String, offset: Int): Int {
        if (plain.isEmpty()) return 1
        val clamped = offset.coerceIn(0, plain.length)
        var count = 1
        for (i in 0 until clamped) if (plain[i] == '\n') count++
        return count
    }

    private const val WT_CLOSE = "</w:t>"
    private val VAL_RE = Regex("w:val=\"([^\"]*)\"")
    private val FILL_RE = Regex("w:fill=\"([0-9A-Fa-f]{6})\"")

    private fun isRealFill(tag: String): Boolean {
        val fill = FILL_RE.find(tag)?.groupValues?.get(1) ?: return false
        return !fill.equals("auto", ignoreCase = true) && !fill.equals("FFFFFF", ignoreCase = true)
    }

    /** A toggle property like `<w:b/>` is ON unless `w:val` says otherwise. */
    private fun toggleOn(tag: String): Boolean {
        val v = VAL_RE.find(tag)?.groupValues?.get(1) ?: return true
        return v != "false" && v != "0" && v != "off" && v != "none"
    }

    /**
     * [styles] is optional: when provided (from [StyleResolver.parse]), paragraph
     * styles (`<w:pStyle>`) and run character styles (`<w:rStyle>`) contribute to
     * the effective bold/italic for each run. When null, only direct `<w:b>`/`<w:i>`
     * properties are used (same as the previous behaviour).
     */
    fun build(xml: String, styles: Map<String, StyleResolver.Props>? = null): PlainMap {
        val sb = StringBuilder()
        val offsets = ArrayList<Int>()
        val formats = ArrayList<FormatSpan>()
        val headings = ArrayList<Heading>()
        val paragraphStyles = ArrayList<ParagraphStyleSpan>()
        // Outline tracking: where the current paragraph's text begins in [sb], and
        // its heading level (null = not a heading), captured from <w:pStyle>.
        var paraStart = 0
        var paraHeadingLevel: Int? = null
        // Paragraph border/shading tracking (only meaningful while inPPr, not inRun).
        var inPBdr = false
        var paraHasBorder = false
        var paraHasShd = false
        // Direct run formatting, tracked inside <w:r>.
        var inRun = false
        var runBold = false
        var runItalic = false
        // Style-based formatting (paragraph level and run character style).
        var inPPr = false   // inside <w:pPr> (only when NOT inside a run)
        var inRPr = false   // inside <w:rPr> (only when inside a run)
        var pStyleBold = false
        var pStyleItalic = false
        var rStyleBold = false
        var rStyleItalic = false

        val n = xml.length
        var i = 0
        while (i < n) {
            val lt = xml.indexOf('<', i)
            if (lt < 0) break
            val gt = xml.indexOf('>', lt)
            if (gt < 0) break

            val tag = xml.substring(lt, gt + 1) // includes '<' and '>'
            val isEnd = tag.startsWith("</")
            val isSelfClose = tag.endsWith("/>")

            // Element name = chars after '<' (or '</') up to whitespace/'/'/'>'.
            val nameStart = if (isEnd) 2 else 1
            var k = nameStart
            while (k < tag.length) {
                val c = tag[k]
                if (c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '/' || c == '>') break
                k++
            }
            val name = tag.substring(nameStart, k)

            if (!isEnd && !isSelfClose && name == "w:t") {
                val contentStart = gt + 1
                val close = xml.indexOf(WT_CLOSE, contentStart)
                val end = if (close < 0) n else close
                val spanStart = sb.length
                // Walk raw XML, decoding entities inline and recording each decoded
                // char's raw start offset. Using decoded-string indices (contentStart+j)
                // would be wrong after any entity whose raw length differs from 1.
                var r = contentStart
                while (r < end) {
                    if (xml[r] == '&') {
                        val semi = xml.indexOf(';', r)
                        if (semi in r + 1 until end + 1) {
                            val entity = xml.substring(r, semi + 1)
                            val entityDecoded = XmlEntities.decode(entity)
                            for (ch in entityDecoded) { sb.append(ch); offsets.add(r) }
                            r = semi + 1
                            continue
                        }
                    }
                    sb.append(xml[r]); offsets.add(r)
                    r++
                }
                val effBold   = runBold   || rStyleBold   || pStyleBold
                val effItalic = runItalic || rStyleItalic || pStyleItalic
                if ((effBold || effItalic) && sb.length > spanStart) {
                    formats.add(FormatSpan(spanStart, sb.length, effBold, effItalic))
                }
                i = if (close < 0) n else close + WT_CLOSE.length
                continue
            }

            when {
                !isEnd && name == "w:tab" -> {
                    sb.append('\t'); offsets.add(lt)
                }
                !isEnd && (name == "w:br" || name == "w:cr") -> {
                    sb.append('\n'); offsets.add(lt)
                }
                // New paragraph: reset paragraph-level style state.
                !isEnd && !isSelfClose && name == "w:p" -> {
                    pStyleBold = false; pStyleItalic = false; inPPr = false
                    paraStart = sb.length; paraHeadingLevel = null
                    inPBdr = false; paraHasBorder = false; paraHasShd = false
                }
                isEnd && name == "w:p" -> {
                    // Record the outline entry from this paragraph's text BEFORE the
                    // trailing newline is appended. Skip blank headings.
                    val lvl = paraHeadingLevel
                    if (lvl != null) {
                        val text = sb.substring(paraStart, sb.length).trim()
                        if (text.isNotEmpty()) headings.add(Heading(text, lvl, paraStart))
                    }
                    if (paraHasBorder || paraHasShd) {
                        paragraphStyles.add(ParagraphStyleSpan(paraStart, sb.length, blockquote = true))
                    }
                    sb.append('\n'); offsets.add(lt)
                }
                // Paragraph properties block (only outside runs).
                !isEnd && !isSelfClose && name == "w:pPr" && !inRun -> inPPr = true
                isEnd && name == "w:pPr" -> inPPr = false
                // Paragraph border block — any edge present counts as a "change bar" style.
                !isEnd && !isSelfClose && name == "w:pBdr" && inPPr -> inPBdr = true
                isEnd && name == "w:pBdr" -> inPBdr = false
                isSelfClose && inPBdr && (name == "w:left" || name == "w:top" || name == "w:bottom" || name == "w:right") ->
                    paraHasBorder = true
                // Paragraph-level shading (distinct from run-level <w:rPr><w:shd> imported as .highlight).
                isSelfClose && name == "w:shd" && inPPr && !inRun && isRealFill(tag) -> paraHasShd = true
                // Paragraph style reference → resolve bold/italic from the style map.
                isSelfClose && name == "w:pStyle" && inPPr && styles != null -> {
                    val sid = VAL_RE.find(tag)?.groupValues?.get(1)
                    val sp = if (sid != null) styles[sid] else null
                    pStyleBold   = sp?.bold   ?: false
                    pStyleItalic = sp?.italic ?: false
                    paraHeadingLevel = sp?.outlineLevel
                }
                // Run open/close.
                !isEnd && !isSelfClose && name == "w:r" -> {
                    inRun = true; runBold = false; runItalic = false
                    rStyleBold = false; rStyleItalic = false; inRPr = false
                }
                isEnd && name == "w:r" -> {
                    inRun = false; runBold = false; runItalic = false
                    rStyleBold = false; rStyleItalic = false; inRPr = false
                }
                // Run properties block (only inside runs).
                !isEnd && !isSelfClose && name == "w:rPr" && inRun -> inRPr = true
                isEnd && name == "w:rPr" && inRun -> inRPr = false
                // Run character style → resolve bold/italic.
                isSelfClose && name == "w:rStyle" && inRun && inRPr && styles != null -> {
                    val sid = VAL_RE.find(tag)?.groupValues?.get(1)
                    val sp = if (sid != null) styles[sid] else null
                    rStyleBold   = sp?.bold   ?: false
                    rStyleItalic = sp?.italic ?: false
                }
                // Direct run bold/italic (override style-based; e.g. <w:b w:val="false"/>).
                !isEnd && inRun && name == "w:b" -> runBold = toggleOn(tag)
                !isEnd && inRun && name == "w:i" -> runItalic = toggleOn(tag)
            }
            i = gt + 1
        }
        return PlainMap(sb.toString(), offsets.toIntArray(), formats, headings, paragraphStyles)
    }
}

File: macos_native/Packages/LeamhDocx/Sources/LeamhDocx/PlainTextMapper.swift
import Foundation

/// A bold and/or italic run over [start, end) char offsets into PlainMap.plain.
public struct FormatSpan: Equatable {
    public let start: Int
    public let end: Int
    public let bold: Bool
    public let italic: Bool
}

/// A whole paragraph over [start, end) char offsets (end excludes the trailing `\n`)
/// styled with a Word paragraph border and/or paragraph-level shading — e.g. a
/// block-quote/callout paragraph (<w:pBdr> + <w:shd> inside <w:pPr>). Distinct from
/// run-level <w:rPr><w:shd> which imports as a .highlight annotation instead (see
/// NativeImport). Mirrors ParagraphStyleSpan in PlainTextMapper.kt.
public struct ParagraphStyleSpan: Equatable {
    public let start: Int
    public let end: Int
    public let blockquote: Bool

    public init(start: Int, end: Int, blockquote: Bool) {
        self.start = start
        self.end = end
        self.blockquote = blockquote
    }
}

/// The canonical plain text plus, for each of its UTF-16 code units, the char offset
/// xmlOffsets into the source document.xml. The two are parallel: xmlOffsets.count == plain.utf16.count.
///
/// formats is a read-only, additive overlay (direct w:b/w:i run formatting) for display only.
public struct PlainMap {
    public let plain: String
    public let xmlOffsets: [Int]
    public let formats: [FormatSpan]
    /// Heading paragraphs in document order — drives the navigation outline.
    public let headings: [Heading]
    /// Block-quote-styled paragraphs in document order — display only, same overlay contract as formats.
    public let paragraphStyles: [ParagraphStyleSpan]

    public init(plain: String, xmlOffsets: [Int], formats: [FormatSpan] = [], headings: [Heading] = [], paragraphStyles: [ParagraphStyleSpan] = []) {
        self.plain = plain
        self.xmlOffsets = xmlOffsets
        self.formats = formats
        self.headings = headings
        self.paragraphStyles = paragraphStyles
    }
}

/// A heading paragraph for the document-outline navigator. level is 0-based
/// (0 = Heading 1). charOffset is the start of the heading's text in PlainMap.plain
/// — divide by plain.utf16.count for the 0.0–1.0 jump fraction (same coordinate
/// system as annotation positions).
///
/// Parity note vs PlainTextMapper.kt: the Swift mapper has no styles.xml
/// resolution, so the heading level is inferred from the conventional
/// `Heading N` pStyle val (Word/Pages/Google Docs all emit this). It does not
/// read an explicit `<w:outlineLvl>` from a custom style. For standard documents
/// the result is identical.
public struct Heading: Equatable {
    public let text: String
    public let level: Int
    public let charOffset: Int

    public init(text: String, level: Int, charOffset: Int) {
        self.text = text
        self.level = level
        self.charOffset = charOffset
    }
}

/// Builds the ONE canonical plain-text string used for both rendering and annotation anchoring.
/// Mirrors PlainTextMapper.kt — clean extraction parsing real element names.
///
/// Extraction rules:
///   <w:t>…</w:t>     decoded text; one xmlOffset per UTF-16 code unit
///   <w:tab/>          '\t'   offset = tag start
///   <w:br/> <w:cr/>   '\n'   offset = tag start
///   </w:p>            '\n'   offset = tag start
///   everything else   ignored
public enum PlainTextMapper {

    /// 1-indexed paragraph number containing UTF-16 offset `offset` into `plain`. Exact —
    /// paragraphs in `plain` are delimited one-for-one by the `\n` each `</w:p>` emits (see
    /// the extraction rules above), so this is a plain newline count, not an approximation.
    /// Mirrors `PlainTextMapper.paragraphIndex` in Kotlin.
    public static func paragraphIndex(_ plain: String, _ offset: Int) -> Int {
        let units = plain.utf16
        guard !units.isEmpty else { return 1 }
        let clamped = max(0, min(offset, units.count))
        var count = 1
        for (i, u) in units.enumerated() {
            if i >= clamped { break }
            if u == 0x0A { count += 1 }
        }
        return count
    }

    private static let wtClose = "</w:t>"
    private static let valRE = try! NSRegularExpression(pattern: #"w:val="([^"]*)""#)
    // Conventional heading styleId "Heading1".."Heading9" (optional space), case-insensitive.
    private static let headingRE = try! NSRegularExpression(pattern: #"^[Hh]eading\s*([1-9])$"#)
    private static let fillRE = try! NSRegularExpression(pattern: #"w:fill="([0-9A-Fa-f]{6})""#)

    private static func isRealFill(_ tag: String) -> Bool {
        guard let m = fillRE.firstMatch(in: tag, range: NSRange(tag.startIndex..., in: tag)) else { return false }
        let fill = (tag as NSString).substring(with: m.range(at: 1))
        return fill.caseInsensitiveCompare("auto") != .orderedSame
            && fill.caseInsensitiveCompare("FFFFFF") != .orderedSame
    }

    /// The 0-based outline level for a pStyle val, or nil if it isn't a heading.
    private static func headingLevel(_ val: String) -> Int? {
        guard let m = headingRE.firstMatch(in: val, range: NSRange(val.startIndex..., in: val)) else { return nil }
        guard let n = Int((val as NSString).substring(with: m.range(at: 1))) else { return nil }
        return n - 1
    }

    private static func toggleOn(_ tag: String) -> Bool {
        guard let m = valRE.firstMatch(in: tag, range: NSRange(tag.startIndex..., in: tag)) else { return true }
        let v = (tag as NSString).substring(with: m.range(at: 1))
        return v != "false" && v != "0" && v != "off" && v != "none"
    }

    public static func build(_ xml: String) -> PlainMap {
        var plain = ""
        var offsets: [Int] = []
        var formats: [FormatSpan] = []
        var headings: [Heading] = []
        var paragraphStyles: [ParagraphStyleSpan] = []
        // Outline tracking: where the current paragraph's text begins in plain (UTF-16),
        // and its heading level (nil = not a heading), captured from <w:pStyle>.
        var paraStart = 0
        var paraHeadingLevel: Int? = nil
        // Paragraph border/shading tracking (only meaningful while !inRun).
        var inPBdr = false
        var paraHasBorder = false
        var paraHasShd = false

        var inRun = false
        var runBold = false
        var runItalic = false

        var i = xml.startIndex

        while i < xml.endIndex {
            guard let ltIdx = xml[i...].firstIndex(of: "<") else { break }
            guard let gtIdx = xml[ltIdx...].firstIndex(of: ">") else { break }

            // UTF-16 offset of tag start (used for xmlOffsets of synthetic chars)
            let tagStartUtf16 = xml.utf16.distance(from: xml.startIndex, to: ltIdx)

            let tag = String(xml[ltIdx...gtIdx]) // includes '<' and '>'
            let isEnd = tag.hasPrefix("</")
            let isSelfClose = tag.hasSuffix("/>")

            // Element name
            let nameStartOff = isEnd ? 2 : 1
            var nameEnd = tag.index(tag.startIndex, offsetBy: nameStartOff)
            while nameEnd < tag.endIndex {
                let c = tag[nameEnd]
                if c == " " || c == "\t" || c == "\n" || c == "\r" || c == "/" || c == ">" { break }
                nameEnd = tag.index(after: nameEnd)
            }
            let name = String(tag[tag.index(tag.startIndex, offsetBy: nameStartOff)..<nameEnd])

            if !isEnd && !isSelfClose && name == "w:t" {
                let contentStart = xml.index(after: gtIdx)
                let contentStartUtf16 = xml.utf16.distance(from: xml.startIndex, to: contentStart)

                if let closeRange = xml.range(of: wtClose, range: contentStart..<xml.endIndex) {
                    let rawContent = String(xml[contentStart..<closeRange.lowerBound])
                    let decoded = XmlEntities.decode(rawContent)
                    let spanStart = plain.utf16.count
                    // One xmlOffset per UTF-16 code unit — mirrors Kotlin's char-by-char loop.
                    // Emoji (2 UTF-16 code units) get 2 consecutive offsets; the string itself
                    // is appended whole so surrogate pairs are preserved correctly.
                    for j in 0..<decoded.utf16.count {
                        offsets.append(contentStartUtf16 + j)
                    }
                    plain += decoded
                    if (runBold || runItalic) && plain.utf16.count > spanStart {
                        formats.append(FormatSpan(start: spanStart, end: plain.utf16.count, bold: runBold, italic: runItalic))
                    }
                    i = closeRange.upperBound
                    continue
                } else {
                    i = xml.endIndex
                    continue
                }
            }

            switch (isEnd, isSelfClose, name) {
            case (false, _, "w:tab"):
                plain += "\t"; offsets.append(tagStartUtf16)
            case (false, _, "w:br"), (false, _, "w:cr"):
                plain += "\n"; offsets.append(tagStartUtf16)
            case (false, false, "w:p"):
                paraStart = plain.utf16.count
                paraHeadingLevel = nil
                inPBdr = false; paraHasBorder = false; paraHasShd = false
            case (true, _, "w:p"):
                // Record the outline entry from this paragraph's text BEFORE the
                // trailing newline is appended. Skip blank headings.
                if let lvl = paraHeadingLevel {
                    let lo = plain.utf16.index(plain.utf16.startIndex, offsetBy: paraStart)
                    let text = String(decoding: Array(plain.utf16[lo...]), as: UTF16.self)
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                    if !text.isEmpty { headings.append(Heading(text: text, level: lvl, charOffset: paraStart)) }
                }
                if paraHasBorder || paraHasShd {
                    paragraphStyles.append(ParagraphStyleSpan(start: paraStart, end: plain.utf16.count, blockquote: true))
                }
                plain += "\n"; offsets.append(tagStartUtf16)
            case (false, _, "w:pStyle") where !inRun:
                if let m = valRE.firstMatch(in: tag, range: NSRange(tag.startIndex..., in: tag)) {
                    paraHeadingLevel = headingLevel((tag as NSString).substring(with: m.range(at: 1)))
                }
            case (false, false, "w:pBdr") where !inRun:
                inPBdr = true
            case (true, _, "w:pBdr"):
                inPBdr = false
            case (false, true, "w:left") where inPBdr:
                paraHasBorder = true
            case (false, true, "w:top") where inPBdr:
                paraHasBorder = true
            case (false, true, "w:bottom") where inPBdr:
                paraHasBorder = true
            case (false, true, "w:right") where inPBdr:
                paraHasBorder = true
            case (false, true, "w:shd") where !inRun && isRealFill(tag):
                paraHasShd = true
            case (false, false, "w:r"):
                inRun = true; runBold = false; runItalic = false
            case (true, _, "w:r"):
                inRun = false; runBold = false; runItalic = false
            case (false, _, "w:b") where inRun:
                runBold = toggleOn(tag)
            case (false, _, "w:i") where inRun:
                runItalic = toggleOn(tag)
            default:
                break
            }

            i = xml.index(after: gtIdx)
        }

        return PlainMap(plain: plain, xmlOffsets: offsets, formats: formats, headings: headings, paragraphStyles: paragraphStyles)
    }
}
